### ISSUE TITLE
Expand delayed/duplicate token handling approach

### ISSUE THREAD

**Issue by guardrex** [aspnet/AspNetCore.Docs#9785]
@cocowalla surfaces an alternative approach to deal with multiple change events. The current approach in the topic+sample uses file hashes (duplicate handling) and an exponential back-off (delay handling). Let's consider revising the current approach or adding the alternate approach. If we add a new approach, we'll add it into the topic (and possibly sample app) over cross-linking. Keep an eye on the issue, as engineering may take it on in a future release.

Let's look at this early next year. We need to get past 2.2 RTM and a few other high priority issues.

* https://github.com/aspnet/AspNetCore/issues/2542#issuecomment-443859286
* https://gist.github.com/cocowalla/5d181b82b9a986c6761585000901d1b8

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9310e951-5185-4c31-064d-3a151acc17cf
* Version Independent ID: 42bc4b7b-2a1c-0248-2740-0decdda6000c
* Content: [Detect changes with change tokens in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/change-tokens?view=aspnetcore-2.1)
* Content Source: [aspnetcore/fundamentals/change-tokens.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/fundamentals/change-tokens.md)
* Product: **aspnet-core**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by Rick-Anderson** [comment]
PV 1.2 K
This is not tracked in #10032

**Comment by guardrex** [comment]
ok on the close, but fyi tho ... #10032 is a different issue and unrelated to this one.

@cocowalla, I'll monitor the engineering topic where the discussion took place. If the engineers make a move, then I'll make a move on the sample. If they call for changes to the topic+sample directly, of course, I'll react to that over here. For now tho, we'll go with what we have (checking file hashes with an exponential backoff).

**Comment by Rick-Anderson** [comment]
> ok on the close, but fyi tho ... #10032 is a different issue and unrelated to this one.

True, but we don't need two issues for low traffic docs.

**Comment by guardrex** [comment]
@cocowalla I wasn't clear. When I wrote, "make a move," I meant that I would **_open a new issue._**

It's not my call if this issue is re-opened or not. It's management's call. I'm just a tiny green dinosaur around here. :smile:

**Comment by Rick-Anderson** [comment]
@cocowalla 	We're not dropping the issue. I just added it to the top of #10032

**Comment by Rick-Anderson** [comment]
Thanks for contacting us.
We don’t have the resources to invest in this area, so we are closing the issue. Should your request generate enough  👍 responses, we’ll reconsider.

**Comment by dazinator** [comment]
Both this and #10032 are closed - but what happened specifically with this issue in the end - was there any progress?

**Comment by guardrex** [comment]
@dazinator Nothing on the docs side. We're still using the same approaches: file hashes (duplicate handling) and an exponential back-off (delay handling). There was discussion https://github.com/dotnet/aspnetcore/issues/2542, but it was closed as a *won't fix*. My approach was deemed "*reasonable*" (that's actually pretty good for my *RexHacks*:tm: around here :smile:). There's no additional info from engineering AFAIK, but you can search their issues to see if someone asked elsewhere. I have no idea at this point what you may find on SO or in chat ... it's been a while since the scenarios were addressed, and I haven't looked again.

**Comment by dazinator** [comment]
@guardrex ok thanks. The reason being is I'm still seeing OptionsMonitor.OnChange() fire twice in quick succession when saving a change to the appsettings.json file at runtime. I wonder if engineering have any tests in place for this scenario -my understandind is that with the new file hash checking logic in place, is should only fire once?

**Comment by guardrex** [comment]
> the new file hash checking logic

If you're referring to something *new* that they did in the framework, perhaps so. You'd have to ask them about that. I'm not aware of any changes on the framework side that pertain to the multiple firing. If I learn more about that, then this topic+sample can receive updates that should shed its workarounds.

If you're referring to what's in the topic ... the file hashes and back-off that I put in there, no, it's a *workaround* for multiple firings. Multiple firing was not **_directly_** addressed by *my hacks*.

**Comment by dazinator** [comment]
@guardrex 
I see. So the workaround is [this one yeah?](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/fundamentals/change-tokens/samples/3.x/SampleApp/Startup.cs#L64)

I think from glancing there might be a possible issue with that hack! From glancing, if `InvokeChanged()` method is fired in parallel, then your if check can be entered into by multiple threads at once (as no locking is used) so your logic could run more than once:

```
 if (!_appsettingsHash.SequenceEqual(appsettingsHash) || 
                !_appsettingsEnvHash.SequenceEqual(appsettingsEnvHash))
            {
                _appsettingsHash = appsettingsHash;
                _appsettingsEnvHash = appsettingsEnvHash;

                WriteConsole("Configuration changed (Simple Startup Change Token)");
            }

```

That aside, may I formally lodge an expression of my discontent over the way this issue has been chucked over the fence from engineering. Without any tests in place, and only a link to a hack, it appears almost like this issue has not really been taken too seriously by engineering, in terms of -  its an issue they are happy to be worked around by everyone as needed when it occurs. If that is the case, the workaround )in this case your hack!) should probably be added as a test on the engineering side so it can be tested more rigorously if this is a solution you are essentially recommending.

I also think that this example of debouncing change tokens based on re-reading the file and computing a hash, may be ok in some situations, but likely not all. It has undesirable characteristics especially when scaling:

1. In a web farm scenario (say scaling on Azure App Service) - if monitoring for a change to a central config file, when that file changes, all nodes in the cluster (which could be very large) will all attempt to re-open the changed file at the same time, and read it to compute a new hash. This places a sudden load on the file system (and accross the network if using network based storage), that could be avoided with other approaches.

2. If the file is quite large, i'm guessing it will take longer to compute the hash. Perhaps too long? https://stackoverflow.com/questions/3837737/improve-performance-of-sha-1-computehash

There is an alternative method that avoids these characteristics, i.e the configurable delay method [as described here.](https://gist.github.com/cocowalla/5d181b82b9a986c6761585000901d1b8)
 but has its own negative:
- Have to configure a sensible delay, if you get it wrong by choosing a timespan that is too small you may not get the benefit of a debounce.

However I can see that both approaches could be used as viable workarounds for this source of errors. Would engineering be prepared to add some tests for a few of these workarounds, and the docs link to those tests instead of your sample code, to proof that these workarounds are "supported" and work?

**Comment by guardrex** [comment]
It's a hack/workaround, and I agree that this one is best in the simplest cases. I felt that I had to do *something* lol. The multiple firings were a pain point. On review when we did this, I didn't discuss the concerns that you raised or any other approaches. They just said that it's *reasonable*, and we moved on. 🏃😅 

If you open a new engineering issue (https://github.com/dotnet/aspnetcore/issues) to discuss further with them:

* Cross-link that original engineering issue at https://github.com/dotnet/aspnetcore/issues/2542 ... and this one + the others perhaps (the SO discussion and possibly the @cocowalla gist at https://gist.github.com/cocowalla/5d181b82b9a986c6761585000901d1b8).
* Please add a `cc: @guardrex` to the bottom of your opening comment so that I can follow the discussion. If they fix something or if they say *nah, let's not use that particular hack*, then I'll open a new issue to address their guidance or track on their issue for their framework fix.

**Comment by dazinator** [comment]
@guardrex done: https://github.com/dotnet/aspnetcore/issues/19401
